labgrid/util/ssh.py

```diff
@@ -0,0 +1,324 @@
+
+    def request_forward(self, host, dest, port):
+        con = self.get(host)
+        return con.add_port_forward(dest, port)
+
+    def remove_forward(self, host, dest, port):
+        con = self.get(host)
+        con.remove_port_forward(dest, port)
+
+    def put_file(self, host, local_file, remote_file):
+        con = self.get(host)
+        con.put_file(local_file, remote_file)
+
+    def get_file(self, host, remote_file, local_file):
+        con = self.get(host)
+        con.get_file(remote_file, local_file)
+
+    def close_all(self):
+        """Close all open connections and remove them from the manager """
+        for name, connection in self._connections.items():
+            if connection.isconnected():
+                connection.disconnect()
+        names = self._connections.copy().keys()
+        for name in names:
+            self.remove_by_name(name)
+
+
+def check_connected(func):
+    """Check if an SSHConnection is connected as a decorator"""
+
+    @wraps(func)
+    def wrapper(cls, *_args, **_kwargs):
+        if not cls.isconnected():
+            raise ExecutionError(
+                "{} can not be called ({} is not connected)".format(
+                    func.__qualname__, cls
+                )
+            )
+        return func(cls, *_args, **_kwargs)
+
+    return wrapper
+
+
+@attr.s
+class SSHConnection:
+    """SSHConnections are individual connections to hosts managed by a control
+    socket. In addition to command execution this class also provides an
+    interface to manage port forwardings. These are used in the remote
+    infrastructure to tunnel multiple connections over one SSH link.
+
+    A public identity infrastructure is assumed, no extra username or passwords
+    are supported."""
+    host = attr.ib(validator=attr.validators.instance_of(str))
+    _connected = attr.ib(
+        default=False, validator=attr.validators.instance_of(bool)
+    )
+    _tmpdir = attr.ib(
+        default=attr.
+        Factory(lambda: tempfile.mkdtemp(prefix="labgrid-connection-")),
+        validator=attr.validators.instance_of(str)
+    )
+    _forwards = attr.ib(default=attr.Factory(dict))
+
+    def __attrs_post_init__(self):
+        self._logger = logging.getLogger("{}".format(self))
+        self._ssh_prefix = "-o LogLevel=ERROR"
+        self._ssh_prefix = "-o PasswordAuthentication=no"
+        self._socket = os.path.join(
+            self._tmpdir, 'control-{}'.format(self.host)
+        )
+
+    def _open_connection(self):
+        """Internal function which appends the control socket and checks if the
+        connection is already open"""
+        self._ssh_prefix += " -o ControlPath={}".format(
+            self._socket
+        ) if self._check_master() else ""
+
+    def _run_socket_command(self, command, forward=""):
+        "Internal function to send a command to the control socket"
+        complete_cmd = "ssh -x -o ControlPath={cpath} -O {command}{forward} {host}".format(
```

Comment:
By using "{command}{forward}", you require the caller to add the space themselves. Just check if a forward is given and then build the command directly as a list, as this also avoids possible problems with spaces in the arguments.

Reply:
I'll rework this and the other commands into lists